### PR TITLE
Check if there is a newer version of diesel in the dependency graph.

### LINKED ISSUE #1704: Yank old versions of diesel_codegen
We seeing quite periodic questions/bugreports about something broken in diesel that boils down to someone tries to use the old `diesel_codegen`. 
I suggest to yank at least the 0.16 version to indicate that this crate shouldn't be used anymore.

### PR DESCRIPTION
**Pull request by weiznich**
This means most likely that the build will break because of a changed codegen.

Fixes #1704 

@sgrif The 0.16.x branch is missing. Could you crate one?

### CODE REVIEW

**Comment by sgrif** [comment]
@weiznich You should have all the rights you need to create it

**Comment by weiznich** [comment]
> You should have all the rights you need to create it.

Right. (I totally forgot about that 😟 )

File: diesel_codegen/src/check_version.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+
```
Comment (Eijebong) [inline diesel_codegen/src/check_version.rs]: :scissors:

**Comment by weiznich** [comment]
> Any reason why you aren't you running this in a build.rs ?

Because I've found no way to get the working directory of a down stream crate from `build.rs`. `CARGO_MANIFEST_DIR` is set there to the directory where `diesel_codegen` lives.

**Comment by sgrif** [comment]
Are we unable to run `cargo metadata` in a build script?

File: diesel_codegen/src/check_version.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+#[derive(Deserialize)]
+struct Metadata {
+    packages: Vec<Package>,
+}
+
+#[derive(Deserialize, Debug)]
+struct Package {
+    name: String,
+    version: String,
+}
```
Comment (sgrif) [inline diesel_codegen/src/check_version.rs]: Newline please :smile:

**Comment by sgrif** [comment]
The `.unwrap`s worry me a bit. I don't want to risk breaking old builds with this. Can we make sure we just return and do nothing if something along the way errors?

File: diesel_codegen/src/check_version.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            .unwrap()
+            .stdout,
+    ).unwrap();
+
+    let data: Metadata = ::serde_json::from_str(&contents).unwrap();
+
+    let diesel = data.packages
+        .into_iter()
+        .filter(|p| p.name == "diesel")
+        .collect::<Vec<_>>();
```
Comment (sgrif) [inline diesel_codegen/src/check_version.rs]: We don't need this `.collect`

File: diesel_codegen/Cargo.toml [hunk 4]
```diff
@@ -17,6 +17,9 @@ dotenv = { version = ">=0.8, <0.11", optional = true, default-features = false }
 diesel = { version = "0.16.0", default-features = false }
 diesel_infer_schema = { version = "0.16.0", default-features = false, optional = true }
 clippy = { optional = true, version = "=0.0.138" }
+serde_json = "1"
+serde = "1"
+serde_derive = "1"
```
Comment (sgrif) [inline diesel_codegen/Cargo.toml]: Can we do `serde = { version = "1.0", features = ["derive"] }` instead?

**Comment by sgrif** [comment]
Can you edit `.travis.yml` to run against this branch in your PR too please?

**Comment by weiznich** [comment]
> Are we unable to run cargo metadata in a build script?

It is possible to run `cargo metadata ` in a build-script, but it is run in the diesel_codegen directory and therefore will result in the dependency list of diesel_codegen. Theoretically we could change the working directory, but I've found no way to get the directory of the downstream crate to run `cargo metadata` there. (Totally possible that I miss something)

**Comment by weiznich** [comment]
> Can you edit .travis.yml to run against this branch in your PR too please?

I've added the branch to the `.travis.yml` but it seems like the test suite is broken on current stable + some dependency is broken on the old nightly.

**Comment by sgrif** [comment]
Ah right, we'll need to cherry pick the fix for that.

**Comment by weiznich** [comment]
I got finally a green CI on this. I've disabled the clippy CI entry because some dependency (pq-sys) does not support that old rustc version.
If there are no objectives I will try to release diesel_codegen as 0.16.1 tomorrow. Do this need a changelog entry? Or a distinct tag?

**Comment by sgrif** [comment]
Yes to both

**Pull request merged**